lib/Modules/ADHooks/ADHooks.psm1

```diff
@@ -1166,6 +1166,98 @@ function Add-ComputerToADGroupDeprecated {
     }
 }
 
+function Get-FSMORoles {
+    <#
+    .SYNOPSIS
+    Retrieves the FSMO role holders from one or more Active Directory domains and forests.
+    .DESCRIPTION
+    Get-FSMORoles uses the Get-ADDomain and Get-ADForest Active Directory cmdlets to determine
+    which domain controller currently holds each of the Active Directory FSMO roles.
+    .PARAMETER DomainName
+    One or more Active Directory domain names.
+    .EXAMPLE
+    Get-Content domainnames.txt | Get-FSMORoles
+    .EXAMPLE
+    Get-FSMORoles -DomainName domain1, domain2
+    #>
+    [CmdletBinding()]
+    param(
+        [Parameter(ValueFromPipeline=$True)]
+        [string[]]$DomainName = $env:USERDOMAIN
+    )
+    BEGIN {
+        Import-Module ActiveDirectory -Cmdlet Get-ADDomain, Get-ADForest -ErrorAction SilentlyContinue
+    }
+    PROCESS {
+        foreach ($domain in $DomainName) {
+            Write-JujuInfo "Querying $domain"
+            Try {
+            $problem = $false
+            $addomain = Get-ADDomain -Identity $domain -ErrorAction Stop
```

Comment:
This should only fail if you are not in a domain. In which case, you can check if you are in a domain at the beginning of the function and return if not. It is not a fatal error, as the AD forest might not have been installed yet. That check is inexpensive:

```
$details = Get-CimInstance Win32_ComputerSystem
$details.PartOfDomain
$details.Domain
```